Bluetooth: host: add support for unregister scanner callback

anchao opened the pull request:
Bluetooth: host: add support for unregister scanner callback

This is a pairing function with bt_le_scan_cb_register()
to used for remove the scanner callback from callback list.

Signed-off-by: chao an <anchao@xiaomi.com>

jhedberg commented on include/bluetooth/bluetooth.h:
```diff
@@ -1714,6 +1714,15 @@ int bt_le_scan_stop(void);
+ *
+ *  Remove the callback structure from the list of scanner callback.
+ *
+ *  @param cb Callback struct. Must point to static memory.
```
I guess we don't have any requirement of this being static memory. It could be dynamically allocated as well. The important thing is that it's a pointer to a struct that was previously added with `bt_le_scan_cb_register()`.

anchao replied:
Done

jhedberg commented on include/bluetooth/bluetooth.h:
```diff
@@ -1714,6 +1714,15 @@ int bt_le_scan_stop(void);
+/**
+ *  @brief Unregister scanner packet callbacks.
+ *
+ *  Remove the callback structure from the list of scanner callback.
```
"list of scanner callbacks"

anchao replied:
Done

jhedberg reviewed (changes requested):
Looks generally fine, just a couple of documentation issues

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -8983,6 +8983,11 @@ void bt_le_scan_cb_register(struct bt_le_scan_cb *cb)
+
+void bt_le_scan_cb_unregister(struct bt_le_scan_cb *cb)
+{
+	sys_slist_find_and_remove(&scan_cbs, &cb->node);
```
Is this actually safe from 2 different context? `SYS_SLIST_FOR_EACH_NODE_SAFE` has this warning ` * This and other SYS_SLIST_*() macros are not thread safe.` I cannot see why `SYS_SLIST_FOR_EACH_CONTAINER_SAFE` would be different.

anchao replied:
Hi @joerchan ,
Sorry for reply late,  
In face, this is not called from 2 different contexts, just considering whether users will delete themselves from the callback call, in order to avoid ambiguity, I will remove the SYS_SLIST_FOR_EACH_CONTAINER_SAFE changes, thanks for review.

joerchan replied:
@anchao Sorry, I think I did not explain it properly. The _SAFE variant makes sense so that you can unregister the callback from the recv callback, I'm not objecting to this.

I'm asking if the cb list needs to be protected by a mutex in addition, because the unregister is an API call and can be called from any context. If it is called from a preemtible thread it might be interrupted in the middle of removing the callback from the list.

trond-snekvik replied:
> If it is called from a preemtible thread

Doesn't a lot of functions in the host API have this problem? I was under the impression that calls from preemptive threads weren't really supported?

@jhedberg is this the case?

joerchan replied:
I suppose that is true.  In that case we can take that as a separate issue considering the entire BT API.

And add back the SYS_SLIST_FOR_EACH_CONTAINER_SAFE change. :)

anchao replied:
Thanks for comment, I brought back the SYS_SLIST_FOR_EACH_CONTAINER_SAFE, Please review again.

anchao committed:
Bluetooth: host: add support for unregister scanner callback

This is a pairing function with bt_le_scan_cb_register()
to used for remove the scanner callback from callback list.

Signed-off-by: chao an <anchao@xiaomi.com>
